build-tools/automation/build.groovy

```diff
@@ -10,12 +10,24 @@ def isPr = false                // Default to CI
 
 def buildTarget = 'jenkins'
 
-def stageWithTimeout(stageName, timeoutValue, timeoutUnit, directory, fatal, Closure body) {
+def stageWithTimeout(stageName, timeoutValue, timeoutUnit, directory, fatal, ctAttempts = 0, Closure body) {
     try {
         stage(stageName) {
-            timeout(time: timeoutValue, unit: timeoutUnit) {
-                dir(directory) {
-                    body()
+            def retryAttempt = 0
+            def waitSecondsBeforeRetry = 15
+            retry(ctAttempts) {     // Retry will always invoke the body at least once for an attempt count of 0 or 1
+                timeout(time: timeoutValue, unit: timeoutUnit) {
+                    dir(directory) {
+                        if (retryAttempt > 0) {
+                            echo "ERROR : Stage ${stageName} failed on try #${retryAttempt}. Waiting ${waitSecondsBeforeRetry} seconds"
```

Comment:
Is `ERROR` appropriate here?  Or should this be `WARNING` or `ALERT` or something that *doesn't* imply that a (non-fatal) error has occurred?

("Error" *usually*, but not necessarily, implies "fatal", so...)

Reply:
Addressed with commit https://github.com/xamarin/xamarin-android/commit/e9763f82a15da97d2b8d15b785f4032216b64377. let's go with WARNING to help avoid confusion with actual ERRORs.
i intentionally made this more jarring so the failure--even though it may pass on retry--is noticed in case its the foreshadowing of an escalating or systemic problem hidden due to the retry. i think what we really need to do here longer term is to log telemetry that we would in turn use to determine whether there's a more pervasive problem.